### ISSUE TITLE
Failed Debug.Assert in SendAsync_SendRequestUsingMethodToEchoServerWithContent_Success on Linux

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#4420]
http://dotnet-ci.cloudapp.net/job/dotnet_corefx_linux_debug_tst_prtest/4225/console
```
14:05:57    System.Net.Http.Functional.Tests.HttpClientHandlerTest.SendAsync_SendRequestUsingMethodToEchoServerWithContent_Success(method: "GET", secureServer: True) [FAIL]
14:05:57       System.Diagnostics.Debug+DebugAssertException : The task must have completed if we're getting called back.
14:05:57       
14:05:57          at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
14:05:57          at System.Environment.get_StackTrace()
14:05:57          at System.Diagnostics.Debug.Assert(Boolean condition, String message, String detailMessage)
14:05:57          at System.Net.Http.CurlHandler.MultiAgent.TransferDataFromRequestStream(IntPtr buffer, Int32 length, EasyRequest easy)
14:05:57          at System.Net.Http.CurlHandler.MultiAgent.CurlSendCallback(IntPtr buffer, UInt64 size, UInt64 nitems, IntPtr context)
14:05:57          at Interop.Http.MultiPerform(SafeCurlMultiHandle multiHandle)
14:05:57          at Interop.Http.MultiPerform(SafeCurlMultiHandle multiHandle)
14:05:57          at System.Net.Http.CurlHandler.MultiAgent.WorkerLoop()
14:05:57          at System.Net.Http.CurlHandler.MultiAgent.<EnsureWorkerIsRunning>b__14(Object s)
14:05:57          at System.Threading.Tasks.Task.Execute()
14:05:57          at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
14:05:57          at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
14:05:57          at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
14:05:57          at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
14:05:57       Stack Trace:
14:05:57             at System.Diagnostics.Debug.UnixDebugLogger.ShowAssertDialog(String stackTrace, String message, String detailMessage)
14:05:57             at System.Net.Http.CurlHandler.MultiAgent.TransferDataFromRequestStream(IntPtr buffer, Int32 length, EasyRequest easy)
14:05:57             at System.Net.Http.CurlHandler.MultiAgent.CurlSendCallback(IntPtr buffer, UInt64 size, UInt64 nitems, IntPtr context)
14:05:57          --- End of stack trace from previous location where exception was thrown ---
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
14:05:57             at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
14:05:57             at System.Net.Http.StreamToStreamCopy.<CopyAsync>d__0.MoveNext()
14:05:57          --- End of stack trace from previous location where exception was thrown ---
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
14:05:57             at System.Net.Http.Functional.Tests.HttpClientHandlerTest.<SendAsync_SendRequestUsingMethodToEchoServerWithContent_Success>d__12e.MoveNext()
14:05:57          --- End of stack trace from previous location where exception was thrown ---
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
14:05:57          --- End of stack trace from previous location where exception was thrown ---
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
14:05:57          --- End of stack trace from previous location where exception was thrown ---
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
14:05:57             at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

**Comment by stephentoub** [comment]
Failed again here:
http://dotnet-ci.cloudapp.net/job/dotnet_corefx_linux_debug_tst_prtest/4231/console

**Comment by stephentoub** [comment]
I'm not positive, but we might have an issue here related to unpausing.  When this functionality was first implemented, we only ever paused/unpaused based on response data.  Later we also added pausing/unpausing based on request data.  I think it's possible some of our asserts (including this one) are stale and need to be changed to accomodate a case where a request and a response for the same operation are in-flight at the same time, in which case we could end up in a situation where we're unpaused prematurely from one of their perspectives.  We'll need to replace the asserts in such cases with special handling that re-pauses, and potentially introduce support for only pausing/unpausing just the request or just the response.  Will need some more investigation.

**Comment by kapilash** [comment]
This was also reported in #4187. I could not repro it locally and was leaning for some intermittent issue at the TestServer side.

**Comment by stephentoub** [comment]
#4187 was different: same test, different failure and cause. I expect the recent change in test server is influencing this, as it's now reproing very frequently as of yesterday.

**Comment by stephentoub** [comment]
Fixed by #4449
